File: 2025/apple/LRUCache.java
// Solution 1: HashMap + Double Linked List
public class LRUCache {
    class Node {
        public int key;
        public int val;
        public Node pre;
        public Node next;
        public Node (int key, int val) {
            this.key = key;
            this.val = val;
            this.pre = null;
            this.next = null;
        }
    }

    Map<Integer, Integer> kv = null;
    Map<Integer, Node> map = null;
    Node tail = null;
    Node head = null;
    int cap = 0;
    /*
    * @param capacity: An integer
    */public LRUCache(int capacity) {
        // do intialization if necessary
        this.kv = new HashMap<>();
        this.map = new HashMap<>();
        tail = new Node(-1, -1);
        head = new Node(-1, -1);
        head.next = tail;
        tail.pre = head;
        this.cap = capacity;
    }

    /*
     * @param key: An integer
     * @return: An integer
     */
    public int get(int key) {
        // write your code here
        if (!kv.containsKey(key)) {
            return -1;
        }
        int val = kv.get(key);
        // remove current node
        Node now = map.get(key);
        now.pre.next = now.next;
        now.next.pre = now.pre;
        moveToTail(now);
        return val;
    }

    /*
     * @param key: An integer
     * @param value: An integer
     * @return: nothing
     */
    public void set(int key, int value) {
        // write your code here
        if (get(key) != -1) {
            kv.put(key, value);
            return;
        }
        // not existing
        Node newone = new Node(key, value);
        kv.put(key, value);
        map.put(key, newone);
        if (kv.size() > this.cap) {
            // Need to remove oldest
            removeFromHead(head.next);
        }
        moveToTail(newone);
    }
    private void removeFromHead(Node now) {
        head.next = now.next;
        now.next.pre = head;
        kv.remove(now.key);
        map.remove(now.key);
    }

    private void moveToTail(Node now) {
        tail.pre.next = now;
        now.pre = tail.pre;
        now.next = tail;
        tail.pre = now;
    }
}

File: LinkedList/lru_cache.py
# TODO
# LRU cache, I am not sure where am I write_docstringdict

class LRUCache:
    """
    @param: capacity: An integer
    """
    class Node:
        def __init__(self, key: int, val: int):
            self.val = val
            self.key = key
            self.pre = None
            self.next = None
        
    def __init__(self, capacity):
        # do intialization if necessary
        self.cap = capacity
        # key is key value is Node
        self.lru_dict = {}
        self.head = self.Node(-1, -1)
        self.tail = self.Node(-1, -1)
        self.head.next = self.tail
        self.tail.pre = self.head
    """
    @param: key: An integer
    @return: An integer
    """
    def get(self, key):
        # write your code here
        if key in self.lru_dict:
            # missing remove myself
            node = self.lru_dict[key]
            self.remove_myself(node)
            self.move_to_tail(node)
            return self.lru_dict.get(key).val
        else:
            return -1

    """
    @param: key: An integer
    @param: value: An integer
    @return: nothing
    """
    def set(self, key, value):
        # write your code here
        if key in self.lru_dict:
            # update key
            node = self.lru_dict[key]
            node.val = value
            self.remove_myself(node)
            self.move_to_tail(node)
            return
        # need to add new entry
        if len(self.lru_dict) == self.cap:
            self.remove_head()
        new_node = self.Node(key, value)
        self.lru_dict[key] = new_node
        self.move_to_tail(self.lru_dict.get(key))
    
    def remove_head(self) -> None:
        temp = self.head.next
        self.head.next = temp.next
        temp.next.pre = self.head
        self.lru_dict.pop(temp.key)
    
    def move_to_tail(self, node) -> None:
        temp = self.tail.pre
        temp.next = node
        node.next = self.tail
        node.pre = temp
    def remove_myself(self, node) -> None:
        node.pre.next = node.next
        node.next.pre = node.pre
    

